File: Python/level2/게임 맵 최단거리.py
def solution(maps):
    answer = 0
    visited = [[0] * len(maps[0]) for i in range(len(maps))]
    visited[0][0] = 1
    n, m = len(maps) - 1, len(maps[0]) - 1
    bfs = [[0, 0, 1]]
    x, y = 0, 0
    while len(bfs) and not (x == n and y == m):
        x, y, count = bfs.pop(0)
        if x - 1 >= 0 and maps[x - 1][y] and not visited[x - 1][y]:
            visited[x - 1][y] = 1
            bfs.append([x - 1, y, count + 1])
        if x + 1 <= n and maps[x + 1][y] and not visited[x + 1][y]:
            visited[x + 1][y] = 1
            bfs.append([x + 1, y, count + 1])
        if y - 1 >= 0 and maps[x][y - 1] and not visited[x][y - 1]:
            visited[x][y - 1] = 1
            bfs.append([x, y - 1, count + 1])
        if y + 1 <= m and maps[x][y + 1] and not visited[x][y + 1]:
            visited[x][y + 1] = 1
            bfs.append([x, y + 1, count + 1])
        answer = count
    if not (x == n and y == m):
        answer = -1
    return answer

File: 게임맵최단거리.cpp
#include <vector>
#include <queue>

using namespace std;

struct pos{
    int y;
    int x;
    pos(int tempy, int tempx) {y = tempy; x = tempx;}
};

int solution(vector<vector<int> > maps)
{
    int answer = -1;
    int n = maps.size();
    int m = maps[0].size();
    
    vector<vector<bool>> visited(n, vector<bool>(m));
    vector<vector<int>> distance(n, vector<int>(m));
    queue<pos> Q;
    
    int moveY[4] = {-1, 0, 1, 0};
    int moveX[4] = {0, 1, 0, -1};
    
    Q.push(pos(0, 0));
    visited[0][0] = true;
    distance[0][0] = 1;
    
    while(!Q.empty()){
        pos p = Q.front();
        Q.pop();
        int nowY = p.y;
        int nowX = p.x;
        
        for(int i = 0; i < 4; i++){
            int nextY = nowY + moveY[i];
            int nextX = nowX + moveX[i];
            
            if(nextY < 0 || nextY >= n || nextX < 0 || nextX >= m) continue;
            if(maps[nextY][nextX] == 0) continue;
            if(visited[nextY][nextX]) continue;
            
            Q.push(pos(nextY, nextX));
            visited[nextY][nextX] = true;
            distance[nextY][nextX] = distance[nowY][nowX] + 1;
        }
    }
    if(visited[n-1][m-1]) answer = distance[n-1][m-1];
    
    return answer;
}